doc/services/binary_descriptors/index.rst

```diff
@@ -0,0 +1,125 @@
+.. _binary_descriptors:
+
+Binary Descriptors
+##################
+
+Binary Descriptors are constant data objects storing information about the binary executable.
+Unlike "regular" constants, binary descriptors are linked to a known offset in the binary, making
+them accesible to other programs, such as a different image running on the same device or a host tool.
+A few examples of constants that would make useful binary descriptors are: kernel version, app version,
+build time, compiler version, environment variables, compiling host name, etc.
+
+Binary descriptors are created by using the ``DEFINE_BINDESC_*`` macros. For example:
+
+.. code-block:: c
+
+   #include <zephyr/bindesc.h>
+
+   DEFINE_BINDESC_STR(my_string, 2, "Hello world!"); // Unique ID is 2
+
+``my_string`` could then be accessed using:
+
+.. code-block:: c
+
+   printk("my_string: %s\n", BINDESC_GET_STR(my_string));
+
+But it could also be retrieved by ``west bindesc``:
+
+.. code-block:: bash
+
+   $ west bindesc custom_search STR 2 build/zephyr/zephyr.bin
+   "Hello world!"
+
+Internals
+*********
+Binary descriptors are implemented with a TLV (tag, length, value) header linked
+to a known offset in the binary image. This offset may vary between architectures,
+but generally the descriptors are linked as close to the beginning of the image as
+possible. In architectures where the image must begin with a vector table (such as
+ARM), the descriptors are linked right after the vector table. The reset vector points
+to the beginning of the text section, which is after the descriptors. In architectures
+where the image must begin with executable code (e.g. x86), a jump instruction is injected at
+the beginning of the image, in order to skip over the binary descriptors, which are right
+after the jump instruction.
+
+Each tag is a 16 bit unsigned integer, where the most significant nibble (4 bits) is the type
+(currently uint, string or bytes), and the rest is the ID. The ID is globally unique to each
+descriptor. For example, the ID of the app version string is ``0x800``, and a string
+is denoted by 0x1, making the app version tag ``0x1800``. The length is a 16 bit
+number equal to the length of the data in bytes. The data is the actual descriptor
+value. All binary descriptor numbers (magic, tags, uints) are laid out in memory
+in the endianness native to the SoC. ``west bindesc`` assumes little endian by default,
+so if the image belongs to a big endian SoC, the appropriate flag should be given to the
+tool.
+
+The binary descriptor header starts with the magic number ``0xb9863e5a7ea46046``. It's followed
```

Comment:
How the magic has been selected?

Reply:
I get a different value though (/jk 😜 )

Reply:
> I get a different value though (/jk 😜 )

I guess we are not doing this ;)
```
#define BINDESC_MAGIC hex(random.randint(0,0xffffffffffffffff)
```

> `hex(random.randint(0,0xffffffffffffffff)`

A little bit disappointed: I thought that you have figured out some magic opcode that is invalid for every architecture and will never be generated by compiler and was wondering what was your approach to get that...